service/src/main/java/io/camunda/service/DocumentServices.java

```diff
@@ -53,18 +57,44 @@ public CompletableFuture<DocumentReferenceResponse> createDocument(
                 storeRecord
                     .instance()
                     .createDocument(storeRequest)
-                    .thenApply(this::handleResponse)
+                    .thenApply(this::requireRightOrThrow)
                     .thenApply(
                         result ->
                             new DocumentReferenceResponse(
                                 result.documentId(), storeRecord.storeId(), result.metadata())));
   }
 
+  /** Will never return a failed future; an Either type is returned instead */
+  public CompletableFuture<List<Either<DocumentErrorResponse, DocumentReferenceResponse>>>
+      createDocumentBatch(final List<DocumentCreateRequest> requests) {
+
+    final List<Either<DocumentErrorResponse, DocumentReferenceResponse>> results =
+        new ArrayList<>();
+
+    final List<CompletableFuture<Void>> futures =
+        requests.stream()
+            .map(
+                request -> {
+                  final var storeRequest =
+                      new DocumentCreationRequest(
+                          request.documentId, request.contentInputStream, request.metadata);
+                  return getDocumentStore(request.storeId)
+                      .thenCompose(
+                          storeRecord -> storeRecord.instance().createDocument(storeRequest))
```

Comment:
Since we are doing batch uploads, it would be better to update the `AwsDocumentStore` and `GcpDocumentStore` to increase the executor threadpool to avoid bottlenecks. They both use `Executors.newSingleThreadExecutor()` at the moment which will make our batch uploads sequential

Reply:
Good point! I added a way to customize the executor service when creating the store. We also use a common thread pool now for all stores.